File: src/protocol/src/main/java/com/kingstondynamics/dragonsong/protocol/message/Message.java
package com.kingstondynamics.dragonsong.protocol.message;

import com.kingstondynamics.dragonsong.protocol.utils.Byteinator;
import com.kingstondynamics.dragonsong.protocol.utils.Concatinator;
import com.kingstondynamics.dragonsong.protocol.utils.Readinator;

public abstract class Message {
    
    private final int opcode;
    private final String auditId;
    
    public Message(int opcode, String auditId) {
        this.opcode = opcode;
        this.auditId = auditId;
    }
    
    public Message(Readinator reader) {
        this.opcode = reader.readInt();
        this.auditId = reader.readAuditId();
    }
    
    public byte[] build() {
        
        byte[] opcodeData = Byteinator.intToBytes(opcode);
        byte[] auditIdData = Byteinator.stringToBytes(auditId);
        
        return Concatinator.ConcatinateByteArrays(opcodeData, auditIdData);
    }
    
    public int getOpcode() {
        return opcode;
    }
    
    public String getAuditId() {
        return auditId;
    }
}

File: sdk/unity/sdk/Assets/KingstonDynamics/DragonSong/Protocol/Messaging/Message.cs
using KingstonDynamics.DragonSong.Protocol.Utils;

namespace KingstonDynamics.DragonSong.Protocol.Messaging
{
    public abstract class Message
    {
        private int TypeCode { get; }
        private string AuditId { get; }

        protected Message(int typeCode, string auditId)
        {
            this.TypeCode = typeCode;
            this.AuditId = auditId;
        }

        protected Message(Readinator reader)
        {
            TypeCode = reader.ReadInt();
            AuditId = reader.ReadAuditId();
        }

        public byte[] Build()
        {
            var b1 = Byteinator.IntToBytes(TypeCode);
            var b2 = Byteinator.StringToBytes(AuditId);

            return Concatinator.ConctatinateByteArrays(b1, b2);
        }
    }
}